R/class-Collections.R

```diff
@@ -0,0 +1,415 @@
+setMethod("Collection", signature("missing", "missing", "missing", "missing"), function(name, data, recordIdColumn, ancestorIdColumns) {
+    new("Collection")
+})
+
+
+##### some helpers for building Collections objects from various data sources #####
+
+findCollectionId <- function(dataColName) {
+    # this is the case where the id columns follow no format and the data columns follow `Name [CollectionId_VariableId]` format (downloads)
+    if (grepl("\\[", dataColName)) {
+        varId <- strsplit(dataColName, "\\[")[[1]][2]
+        collectionId <- regmatches(varId,regexpr("^([^_]*_[^_]*)",varId))
+        return(collectionId)
+    }
+
+    # this presumably the case where the column headers follow the `entityId.variableId` format (the eda services format)
+    if (grepl(".", dataColName, fixed = TRUE)) {
+        varId <- strsplit(dataColName, ".", fixed = TRUE)[[1]][1]
+        collectionId <- strsplit(varId, "_", fixed = TRUE)[[1]][1]
+        return(collectionId)
+    }
+
+    stop((sprintf("Could not find collection id for column: %s. Unrecognized format.", dataColName)))
+}
+
+findCollectionIds <- function(dataColNames) {
+    recordIdColumn <- findRecordIdColumn(dataColNames)
+    ancestorIdColumns <- findAncestorIdColumns(dataColNames)
+    variableColNames <- dataColNames[!dataColNames %in% c(recordIdColumn,ancestorIdColumns)]
+
+    return(unique(unlist(sapply(variableColNames, findCollectionId))))
+}
+
+findRecordIdColumn <- function(dataColNames) {
+    # for now assume were working w bulk download files, which means its the first column
+    allIdColumns <- dataColNames[grepl("_Id", dataColNames, fixed=TRUE)]
+    return(allIdColumns[1])
+}
+
+findAncestorIdColumns <- function(dataColNames) {
+    # for now assume were working w bulk download files, which means they have '_Id'
+    allIdColumns <- dataColNames[grepl("_Id", dataColNames, fixed=TRUE)]
+    if (length(allIdColumns) == 1) {
+        return(character(0))
+    }
+
+    return(allIdColumns[2:length(allIdColumns)])
+}
+
+clean_names <- function(names, makeUnique = FALSE) {
+    # remove everything after the last opening square bracket to get rid of IRIs
+    names <- gsub("\\[.*$", "", names)
+
+    names <- gsub("%+", "_pct_", names)
+    names <- gsub("\\$+", "_dollars_", names)
+    names <- gsub("\\++", "_plus_", names)
+    names <- gsub("-+", "_minus_", names)
+    names <- gsub("\\*+", "_star_", names)
+    names <- gsub("#+", "_cnt_", names)
+    names <- gsub("&+", "_and_", names)
+    names <- gsub("@+", "_at_", names)
+
+    names <- gsub("[^a-zA-Z0-9_]+", "_", names)
+    names <- gsub("([A-Z][a-z])", "_\\1", names)
+    names <- tolower(trimws(names))
+
+    names <- gsub("(^_+|_+$)", "", names)
+
+    names <- gsub("_+", "_", names)
+
+    if (makeUnique) names <- make.unique(names, sep = "_")
+
+    return(names)
+}
+
+getDataFromSource <- function(dataSource, keepIdsAndNumbersOnly = c(TRUE, FALSE), cleanColumnNames = c(FALSE, TRUE)) {
```

Comment:
This one could use some docs. Param defs would be helpful, but even if it just says 'hey this is helpful for making the mbio packages' then that would be helpful, too. It looks like something a user would want to use

Reply:
yea, i also considered that it should maybe live somewhere else. but at the end of the day its not quite mbio specific. its useful for any study w collections of assay data on some entities and regular vars on other entities. thatll likely be genomics too, as they get further into eda territory. 

i dont think i want roxygen docs on this, but im happy enough to add some stronger commenting for our future selves or any curious souls who happen along.